src/main/kotlin/com/hypto/iam/server/service/UsersService.kt

```diff
@@ -538,6 +554,136 @@ class UsersServiceImpl : KoinComponent, UsersService {
+            "SUBORDINATE" -> listSubordinateUsers(principal, context)
+            else -> throw IllegalArgumentException("Invalid value found for query param role")
+        }
+    }
+
+    private suspend fun listLeaderUsers(
+        principal: UserPrincipal,
+        context: PaginationContext,
+    ): LinkUsersPaginatedResponse {
+        val linkUserRecords = linkUsersRepo.fetchLeaderUsers(principal.hrnStr, context)
+        val userRecordsMap = userRepo.findByHrns(linkUserRecords.keys.toList())
+        val newContext = PaginationContext.from(linkUserRecords.entries.lastOrNull()?.value?.leaderUser, context)
+        return LinkUsersPaginatedResponse(
+            data =
+                userRecordsMap.map {
+                    val linkUserRecord = linkUserRecords[it.key]!!
+                    val userHrn = ResourceHrn(it.key)
+                    LinkUserResponse(
+                        id = linkUserRecord.id!!,
+                        name = it.value.name ?: "",
+                        organizationId = userHrn.organization,
+                        email = it.value.email,
+                    )
+                },
+            nextToken = newContext.nextToken,
+            context = newContext.toOptions(),
+        )
+    }
+
+    private suspend fun listSubordinateUsers(
+        principal: UserPrincipal,
+        context: PaginationContext,
+    ): LinkUsersPaginatedResponse {
+        val linkUserRecords = linkUsersRepo.fetchSubordinateUsers(principal.hrnStr, context)
+        val userRecordsMap = userRepo.findByHrns(linkUserRecords.keys.toList())
+        val newContext = PaginationContext.from(linkUserRecords.entries.lastOrNull()?.value?.subordinateUser, context)
+        return LinkUsersPaginatedResponse(
+            data =
+                userRecordsMap.map {
+                    val linkUserRecord = linkUserRecords[it.key]!!
+                    val userHrn = ResourceHrn(it.key)
+                    LinkUserResponse(
+                        id = linkUserRecord.id!!,
+                        name = it.value.name ?: "",
+                        organizationId = userHrn.organization,
+                        email = it.value.email,
+                    )
+                },
+            nextToken = newContext.nextToken,
+            context = newContext.toOptions(),
+        )
+    }
+
+    override suspend fun switchUser(
+        principal: UserPrincipal,
+        linkId: String,
+    ): TokenResponse {
+        val leaderUserHrn = principal.claims?.get(ON_BEHALF_CLAIM) as String? ?: principal.hrnStr
+        val record =
+            linkUsersRepo.getById(linkId)
+                ?: throw EntityNotFoundException("Invalid user link id")
+        require(record.leaderUser == leaderUserHrn) {
+            "User doesn't have permission to switch"
+        }
+        return if (record.leaderUser == principal.hrnStr) {
+            tokenService.generateJwtToken(ResourceHrn(record.subordinateUser), principal.hrn)
```

Comment:
Looks like a subordinate user can get access for different subordinate user
Is this acceptable?